src/Middleware/HttpLogging/src/HttpRequestLog.cs

```diff
@@ -37,8 +37,7 @@ public override string ToString()
         {
             if (_cachedToString == null)
             {
-                // TODO use string.Create instead of a StringBuilder here.
-                var builder = new StringBuilder();
+                var builder = new ValueStringBuilder();
```

Comment:
Can we come up with a rough guess of how big the string will be? Otherwise this will need to do a lot of internal copying and renting whenever it grows. For example, each `_keyValues` will have at minimum ~5 characters output + 8 for "Request:\n".

I'd love to see perf numbers from before and after, and with a guess at the size here so we can see that improvements are actually being made.

Reply:
What should I use to get the perf numbers? BenchmarkDotNet?

Reply:
Yeah, take a look at https://github.com/dotnet/aspnetcore/blob/main/src/Middleware/perf/Microbenchmarks/HandshakeBenchmark.cs for an example in the middleware area. You go to that folder and run `dotnet run -c Release` then select which benchmark to run if there are multiple to choose from.

Reply:
Here's the result with this list 
```
list = new List<KeyValuePair<string, string>>();
list.Add(new KeyValuePair<string, string>("Protocol", "HTTP/1.1"));
list.Add(new KeyValuePair<string, string>("Method", "GET"));
list.Add(new KeyValuePair<string, string>("Scheme", "http"));
list.Add(new KeyValuePair<string, string>("PathBase", "/"));
list.Add(new KeyValuePair<string, string>("Path", "/"));
list.Add(new KeyValuePair<string, string>("QueryString", ""));
list.Add(new KeyValuePair<string, string>("Connection", "KeepAlive"));
list.Add(new KeyValuePair<string, string>("Content-Type", "text/plain"));
```
and the initial capacity is `8 + _keyValues.Count * 5`.

|                         Method |     Mean |    Error |   StdDev |   Median |        Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------- |---------:|---------:|---------:|---------:|------------:|-------:|------:|------:|----------:|
|                  StringBuilder | 681.8 ns | 13.62 ns | 32.38 ns | 669.2 ns | 1,466,662.8 | 0.0448 |     - |     - |   1,208 B |
|             ValueStringBuilder | 674.5 ns |  6.26 ns |  5.22 ns | 672.9 ns | 1,482,670.6 | 0.0124 |     - |     - |     336 B |
|      StringBuilderWithCapacity | 530.5 ns |  5.89 ns |  4.60 ns | 529.6 ns | 1,884,897.5 | 0.0343 |     - |     - |     936 B |
| ValueStringBuilderWithCapacity | 684.1 ns |  4.40 ns |  3.68 ns | 683.1 ns | 1,461,677.9 | 0.0124 |     - |     - |     336 B |

Reply:
That's a big difference between `StringBuilderWithCapacity` and `ValueStringBuilderWithCapacity`. Could you try initial capacity of `8 + _keyValues.Count * 100`? In the `* 5` case, none of the names are going to be 1 byte, that was just the bare minimum, so a grow will happen which causes a copy and new rent.

Reply:
Here's the result with the same list and capacity of `8 + _keyValues.Count * 100` :

|                         Method |     Mean |    Error |   StdDev |   Median |        Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------- |---------:|---------:|---------:|---------:|------------:|-------:|------:|------:|----------:|
|                  StringBuilder | 666.4 ns | 11.91 ns | 24.06 ns | 656.8 ns | 1,500,658.5 | 0.0439 |     - |     - |   1,208 B |
|             ValueStringBuilder | 628.7 ns |  9.39 ns |  8.79 ns | 627.6 ns | 1,590,498.3 | 0.0124 |     - |     - |     336 B |
|      StringBuilderWithCapacity | 581.2 ns |  6.66 ns |  6.23 ns | 580.8 ns | 1,720,660.1 | 0.0744 |     - |     - |   2,024 B |
| ValueStringBuilderWithCapacity | 626.6 ns |  2.82 ns |  2.20 ns | 626.6 ns | 1,595,815.1 | 0.0124 |     - |     - |     336 B |

Reply:
Crazy, I'd love to know why the ValueStringBuilder is so slow

Reply:
Oh, sorry, my bad! I forgot to actually pass the capacity to the `ValueStringBuilder` ctor in the benchmark 🤦‍♂️.

This result makes much more sence:

|                         Method |     Mean |    Error |   StdDev |        Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------- |---------:|---------:|---------:|------------:|-------:|------:|------:|----------:|
|                  StringBuilder | 714.2 ns | 20.14 ns | 59.08 ns | 1,400,154.0 | 0.0448 |     - |     - |   1,208 B |
|             ValueStringBuilder | 614.0 ns | 11.14 ns | 10.42 ns | 1,628,775.8 | 0.0124 |     - |     - |     336 B |
|      StringBuilderWithCapacity | 618.4 ns | 21.35 ns | 60.56 ns | 1,617,180.6 | 0.0744 |     - |     - |   2,024 B |
| ValueStringBuilderWithCapacity | 435.0 ns |  6.17 ns |  5.78 ns | 2,298,704.4 | 0.0124 |     - |     - |     336 B |

Reply:
Awesome, thanks! Now we should figure out a reasonable capacity, I assume 100 is a bit high.

Reply:
Are the benchmarks you ran in a good enough state that we could include them in the PR for future usage?

Reply:
I don't think it is good enough. I had to make the `HttpRequestLog` public and also add a method which get the capacity as input. Also I had both `StringBuilder` and `ValueStringBuilder` in `HttpRequestLog` for the benchmark.

Reply:
@Tratcher Thoughts on a good default capacity? We would prefer slightly over-estimating the total size of the headers.

Reply:
All request headers on average? 4-16kb. Start at 2kb for the simple case though. Let the big requests pay the cost.

Reply:
@Tratcher What's the average on all response headers? Is 2kb good enough for response too?

Reply:
That should be fine for now, there tend to be less response headers than request headers.